cpp/src/io/csv/datetime.cuh

```diff
@@ -355,96 +315,136 @@ __inline__ __device__ int64_t parseDateTimeFormat(char const* begin, char const*
+  *begin = thrust::find_if(thrust::seq, *begin, end, [](auto elem) { return elem != ' '; });
 }
 
+/**
+ * @brief Parses the input string into a duration of the given type.
+ *
+ * @param begin Pointer to the first element of the string
+ * @param end Pointer to the last element of the string
+ * @return The parsed duration
+ */
 template <typename T>
-__inline__ __device__ int64_t parseTimeDeltaFormat(const char* data, long start, long end)
+__inline__ __device__ int64_t to_time_delta(char const* begin, char const* end)
 {
+  ++end;
   // %d days [+]%H:%M:%S.n => %d days, %d days [+]%H:%M:%S,  %H:%M:%S.n, %H:%M:%S, %value.
-  int days{0};
-  int8_t hour{0}, minute{0}, second{0};
-  int nanosecond     = 0;
   constexpr char sep = ':';
 
+  int32_t days{0};
+  int8_t hour{0};
   // single pass to parse days, hour, minute, seconds, nanosecond
-  long moving_pos = start;
-  int32_t value   = parse_integer<int>(data, moving_pos, end);
-  while (data[moving_pos] == ' ' && moving_pos <= end) moving_pos++;
-  if (std::is_same<T, cudf::duration_D>::value || moving_pos >= end) {  // %value
+  auto cur         = begin;
+  auto const value = parse_integer<int32_t>(&cur, end);
+  skip_spaces(&cur, end + 1);
+  if (std::is_same<T, cudf::duration_D>::value || cur >= end) {  // %value
     return value;
+  }
+  // " days [+]"
+  bool const has_days_seperator = skip_if_starts_with(&cur, end + 1, "days");
+  skip_spaces(&cur, end + 1);
+  cur += (*cur == '+');
+  if (has_days_seperator) {
+    days = value;
+    hour = parse_integer<int8_t>(&cur, end);
   } else {
-    // " days [+]"
-    const bool days_seperator = is_present(data, moving_pos, end, "days", 4);
-    while (data[moving_pos] == ' ' && moving_pos <= end) moving_pos++;
-    moving_pos += (data[moving_pos] == '+');
-    if (days_seperator) {
-      days = value;
-      hour = parse_integer<int>(data, moving_pos, end);
-    } else {
-      hour = value;
-    }
+    hour = value;
+  }
 
-    //:%M:%S
-    if (data[moving_pos] == sep) { minute = parse_integer<int>(data, ++moving_pos, end); }
-    if (data[moving_pos] == sep) { second = parse_integer<int>(data, ++moving_pos, end); }
-    if (std::is_same<T, cudf::duration_s>::value) {
-      return ((days * 24L + hour) * 60L + minute) * 60L + second;
-    } else if (data[moving_pos] == '.') {  //.n
-      auto start_subsecond              = moving_pos + 1;
-      nanosecond                        = parse_integer<int>(data, ++moving_pos, end);
-      int8_t num_digits                 = min(9L, moving_pos - start_subsecond);
-      constexpr int64_t powers_of_ten[] = {
-        1L, 10L, 100L, 1000L, 10000L, 100000L, 1000000L, 10000000L, 100000000L, 1000000000L};
-      nanosecond *= powers_of_ten[9 - num_digits];
-    }
+  int8_t minute{0};
+  //:%M:%S
+  if (*cur == sep) {
+    ++cur;
+    minute = parse_integer<int8_t>(&cur, end);
+  }
+  int8_t second{0};
+  if (*cur == sep) {
+    ++cur;
+    second = parse_integer<int8_t>(&cur, end);
+  }
```

Comment:
It'd be cool if we could declare and init simultaneously, but this isn't super clean either. :/
```suggestion
  //:%M:%S
  auto minute = parse_optional_integer<int8_t>(&cur, end, sep);
  auto second = parse_optional_integer<int8_t>(&cur, end, sep);
```
```c++
template<typename T>
parse_optional_integer(char** cur, char* end, char sep) {
  if (**cur == sep) {
    (*cur)++;
    return parse_integer<T>(cur, end);
  }
  return 0;
}
```